[BUG] Failed to upload index.html file to web storage account during Azure TRE deployment

Failed to upload index.html file to web storage account during Azure TRE deployment as part of "make letsencrypt" step due to missing required permissions.

**Steps to reproduce**

1. Deploy AzureTRE following this doc - [https://github.com/microsoft/AzureTRE/blob/develop/docs/deployment-quickstart.md](https://github.com/microsoft/AzureTRE/blob/develop/docs/deployment-quickstart.md)
1. When you run ```/workspaces/tre> make letsencrypt```  you should get the following error:


```
vscode ➜ /workspaces/tre (quickstartenv ✗) $ make letsencrypt
echo -e "\n\e[34m»»» 🧩 \e[96mRequesting LetsEncrypt SSL certificate\e[0m..." \
&& . ./devops/scripts/check_dependencies.sh nodocker,certbot \
&& . ./devops/scripts/load_env.sh ./templates/core/.env \
&& . ./devops/scripts/load_env.sh ./devops/.env \
&& . ./devops/scripts/load_terraform_env.sh ./devops/.env \
&& . ./devops/scripts/load_terraform_env.sh ./templates/core/.env \
&& cd ./templates/core/terraform/ && . ./outputs.sh \
&& cd ./scripts/ && ./letsencrypt.sh

»»» 🧩 Requesting LetsEncrypt SSL certificate...

╔══════════════════════════════════════╗
║          Azure TRE Makefile          ║
╚══════════════════════════════════════╝

»»» ✅ Checking pre-reqs...

Checking for Azure CLI...

Checking for Certbot...
»»» 🔨 Azure details from logged on user 
»»»   • Subscription: TamirK-CSE-Israel
»»»   • Tenant:       72f988bf-86f1-41af-91ab-2d7cd011db47

Upgrading modules...
- api-webapp in api-webapp
- appgateway in appgateway
- bastion in bastion
- firewall in firewall
- gitea in ../../shared_services/gitea/terraform
- identity in user-assigned-identity
- keyvault in keyvault
- network in network
- nexus in ../../shared_services/sonatype-nexus/terraform
- resource_processor_vmss_porter in resource_processor/vmss_porter
- routetable in routetable
- servicebus in servicebus
- state-store in state-store
- storage in storage

Initializing the backend...

Successfully configured the backend "azurerm"! Terraform will automatically
use this backend unless the backend configuration changes.

Initializing provider plugins...
- Finding hashicorp/azurerm versions matching "2.71.0"...
- Finding latest version of hashicorp/random...
- Finding latest version of hashicorp/null...
- Finding latest version of hashicorp/template...
- Using previously-installed hashicorp/null v3.1.0
- Using previously-installed hashicorp/template v2.2.0
- Using previously-installed hashicorp/azurerm v2.71.0
- Using previously-installed hashicorp/random v3.1.0

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
RESOURCE_GROUP=rg-gb1-tre
APPLICATION_GATEWAY=agw-gb1-tre
STORAGE_ACCOUNT=stwebgb1tre
KEYVAULT=kv-gb1-tre
FQDN=gb1-tre.westeurope.cloudapp.azure.com
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    14  100    14    0     0     52      0 --:--:-- --:--:-- --:--:--    52
Creating network rule on storage account stwebgb1tre for 77.125.241.247
{
  "accessTier": "Hot",
  "allowBlobPublicAccess": false,
  "allowCrossTenantReplication": null,
  "allowSharedKeyAccess": true,
  "azureFilesIdentityBasedAuthentication": null,
  "blobRestoreStatus": null,
  "creationTime": "2021-09-09T12:24:31.446701+00:00",
  "customDomain": null,
  "enableHttpsTrafficOnly": true,
  "enableNfsV3": false,
  "encryption": {
    "encryptionIdentity": null,
    "keySource": "Microsoft.Storage",
    "keyVaultProperties": null,
    "requireInfrastructureEncryption": null,
    "services": {
      "blob": {
        "enabled": true,
        "keyType": "Account",
        "lastEnabledTime": "2021-09-09T12:24:31.540466+00:00"
      },
      "file": {
        "enabled": true,
        "keyType": "Account",
        "lastEnabledTime": "2021-09-09T12:24:31.540466+00:00"
      },
      "queue": null,
      "table": null
    }
  },
  "extendedLocation": null,
  "failoverInProgress": null,
  "geoReplicationStats": null,
  "id": "/subscriptions/73a4ea93-d914-424d-9e64-28adf397e8e3/resourceGroups/rg-gb1-tre/providers/Microsoft.Storage/storageAccounts/stwebgb1tre",
  "identity": {
    "principalId": null,
    "tenantId": null,
    "type": "None",
    "userAssignedIdentities": null
  },
  "isHnsEnabled": false,
  "keyCreationTime": {
    "key1": "2021-09-09T12:24:31.540466+00:00",
    "key2": "2021-09-09T12:24:31.540466+00:00"
  },
  "keyPolicy": null,
  "kind": "StorageV2",
  "largeFileSharesState": null,
  "lastGeoFailoverTime": null,
  "location": "westeurope",
  "minimumTlsVersion": "TLS1_0",
  "name": "stwebgb1tre",
  "networkRuleSet": {
    "bypass": "AzureServices",
    "defaultAction": "Deny",
    "ipRules": [
      {
        "ipAddressOrRange": "77.125.241.247"
      }
    ],
    "resourceAccessRules": [],
    "virtualNetworkRules": []
  },
  "primaryEndpoints": {
    "blob": "https://stwebgb1tre.blob.core.windows.net/",
    "dfs": "https://stwebgb1tre.dfs.core.windows.net/",
    "file": "https://stwebgb1tre.file.core.windows.net/",
    "internetEndpoints": null,
    "microsoftEndpoints": null,
    "queue": "https://stwebgb1tre.queue.core.windows.net/",
    "table": "https://stwebgb1tre.table.core.windows.net/",
    "web": "https://stwebgb1tre.z6.web.core.windows.net/"
  },
  "primaryLocation": "westeurope",
  "privateEndpointConnections": [
    {
      "id": "/subscriptions/73a4ea93-d914-424d-9e64-28adf397e8e3/resourceGroups/rg-gb1-tre/providers/Microsoft.Storage/storageAccounts/stwebgb1tre/privateEndpointConnections/stwebgb1tre.9364e8b1-0c75-4ac8-b23b-0a71462ce352",
      "name": "stwebgb1tre.9364e8b1-0c75-4ac8-b23b-0a71462ce352",
      "privateEndpoint": {
        "id": "/subscriptions/73a4ea93-d914-424d-9e64-28adf397e8e3/resourceGroups/rg-gb1-tre/providers/Microsoft.Network/privateEndpoints/pe-web-stwebgb1tre",
        "resourceGroup": "rg-gb1-tre"
      },
      "privateLinkServiceConnectionState": {
        "actionRequired": "None",
        "description": "Auto-Approved",
        "status": "Approved"
      },
      "provisioningState": "Succeeded",
      "resourceGroup": "rg-gb1-tre",
      "type": "Microsoft.Storage/storageAccounts/privateEndpointConnections"
    }
  ],
  "provisioningState": "Succeeded",
  "resourceGroup": "rg-gb1-tre",
  "routingPreference": null,
  "sasPolicy": null,
  "secondaryEndpoints": null,
  "secondaryLocation": null,
  "sku": {
    "name": "Standard_LRS",
    "tier": "Standard"
  },
  "statusOfPrimary": "available",
  "statusOfSecondary": null,
  "tags": {
    "tre_id": "gb1-tre"
  },
  "type": "Microsoft.Storage/storageAccounts"
}
Waiting for network rule to take effect
Created network rule on storage account
Static Web hosting already enabled on storage account
Checking for index.html file in storage account
ERROR: 
You do not have the required permissions needed to perform this operation.
Depending on your operation, you may need to be assigned one of the following roles:
    "Storage Blob Data Contributor"
    "Storage Blob Data Reader"
    "Storage Queue Data Contributor"
    "Storage Queue Data Reader"

If you want to use the old authentication method and allow querying for the right account key, please use the "--auth-mode" parameter and "key" value.
                    
Uploading index.html file
Client-Request-ID=0c689f70-1393-11ec-82bf-0242ac120002 Retry policy did not allow for a retry: Server-Timestamp=Sun, 12 Sep 2021 06:31:21 GMT, Server-Request-ID=8a783964-f01e-005b-339f-a7aac1000000, HTTP status code=403, Exception=This request is not authorized to perform this operation using this permission. ErrorCode: AuthorizationPermissionMismatch<?xml version="1.0" encoding="utf-8"?><Error><Code>AuthorizationPermissionMismatch</Code><Message>This request is not authorized to perform this operation using this permission.RequestId:8a783964-f01e-005b-339f-a7aac1000000Time:2021-09-12T06:31:22.4886388Z</Message></Error>.

You do not have the required permissions needed to perform this operation.
Depending on your operation, you may need to be assigned one of the following roles:
    "Storage Blob Data Contributor"
    "Storage Blob Data Reader"
    "Storage Queue Data Contributor"
    "Storage Queue Data Reader"

If you want to use the old authentication method and allow querying for the right account key, please use the "--auth-mode" parameter and "key" value.
                    
make: *** [Makefile:83: letsencrypt] Error 1
```

--- a/templates/core/terraform/appgateway/locals.tf
+++ b/templates/core/terraform/appgateway/locals.tf
@@ -1,6 +1,8 @@
 locals {
   staticweb_storage_name = lower(replace("stweb${var.tre_id}", "-", ""))
 
+  staticweb_index_file_content = "<!DOCTYPE html><html lang=\"en\" xmlns=\"http://www.w3.org/1999/xhtml\"><head><meta charset=\"utf-8\"/><title></title></head><body></body></html>"
+
   staticweb_backend_pool_name = "beap-staticweb"
   api_backend_pool_name       = "beap-api"
   nexus_backend_pool_name     = "beap-nexus"

--- a/templates/core/terraform/appgateway/locals.tf
+++ b/templates/core/terraform/appgateway/locals.tf
@@ -1,8 +1,6 @@
 locals {
   staticweb_storage_name = lower(replace("stweb${var.tre_id}", "-", ""))
 
-  staticweb_index_file_content = "<!DOCTYPE html><html lang=\"en\" xmlns=\"http://www.w3.org/1999/xhtml\"><head><meta charset=\"utf-8\"/><title></title></head><body></body></html>"
-
   staticweb_backend_pool_name = "beap-staticweb"
   api_backend_pool_name       = "beap-api"
   nexus_backend_pool_name     = "beap-nexus"

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -10,21 +10,16 @@ resource "azurerm_storage_account" "staticweb" {
   enable_https_traffic_only = true
   allow_blob_public_access  = false
 
+  tags = {
+    tre_id = var.tre_id
+  }
+
   static_website {
     index_document     = "index.html"
     error_404_document = "404.html"
   }
 
-  tags = {
-    tre_id = var.tre_id
-  }
-
   lifecycle { ignore_changes = [tags] }
-
-  network_rules {
-    bypass         = ["AzureServices"]
-    default_action = "Deny"
-  }
 }
 
 resource "azurerm_storage_container" "staticweb" {
@@ -42,12 +37,16 @@ resource "azurerm_storage_blob" "staticweb" {
   source_content         = local.staticweb_index_file_content
 }
 
-# Assign the identity deploying data contibutor rights.
-# Needed to upload content to static web.
-resource "azurerm_role_assignment" "stgwriter" {
-  scope                = azurerm_storage_account.staticweb.id
-  role_definition_name = "Storage Blob Data Contributor"
-  principal_id         = data.azurerm_client_config.deployer.object_id
+resource "azurerm_storage_account_network_rules" "staticweb" {
+  resource_group_name  = var.resource_group_name
+  storage_account_name = azurerm_storage_account.staticweb.name
+
+  default_action = "Deny"
+  bypass         = ["AzureServices"]
+
+  depends_on = [
+    azurerm_storage_blob.staticweb
+  ]
 }
 
 resource "azurerm_private_endpoint" "webpe" {

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -43,6 +43,14 @@ resource "azurerm_storage_account_network_rules" "staticweb" {
   ]
 }
 
+# Assign the role to allow uploading content to the storage account
+# Needed for uploading certificates
+resource "azurerm_role_assignment" "stgwriter" {
+  scope                = azurerm_storage_account.staticweb.id
+  role_definition_name = "Storage Blob Data Contributor"
+  principal_id         = data.azurerm_client_config.deployer.object_id
+}
+
 resource "azurerm_private_endpoint" "webpe" {
   name                = "pe-web-${local.staticweb_storage_name}"
   location            = var.location

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -9,10 +9,12 @@ resource "azurerm_storage_account" "staticweb" {
   account_replication_type  = "LRS"
   enable_https_traffic_only = true
   allow_blob_public_access  = false
+
   static_website {
     index_document     = "index.html"
     error_404_document = "404.html"
   }
+
   tags = {
     tre_id = var.tre_id
   }
@@ -23,6 +25,20 @@ resource "azurerm_storage_account" "staticweb" {
     bypass         = ["AzureServices"]
     default_action = "Deny"
   }
+}
+
+resource "azurerm_storage_container" "staticweb" {
+  name                  = "$web"
+  storage_account_name  = azurerm_storage_account.staticweb.name
+  container_access_type = "private"
+}
+
+resource "azurerm_storage_blob" "staticweb" {
+  name                   = "index.html"
+  storage_account_name   = azurerm_storage_account.staticweb.name
+  storage_container_name = azurerm_storage_container.staticweb.name
+  type                   = "Block"
+  source_content         = local.staticweb_index_file_content
 }
 
 # Assign the identity deploying data contibutor rights.

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -29,6 +29,8 @@ resource "azurerm_storage_blob" "staticweb" {
   type                   = "Block"
   content_type           = "text/html"
   source_content         = local.staticweb_index_file_content
+
+  lifecycle { ignore_changes = [tags] }
 }
 
 resource "azurerm_storage_account_network_rules" "staticweb" {

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -30,11 +30,9 @@ resource "azurerm_storage_blob" "staticweb" {
   content_type           = "text/html"
   source_content         = local.staticweb_index_file_content
 
-  tags = {
-    tre_id = var.tre_id
+  lifecycle {
+    ignore_changes = all
   }
-
-  lifecycle { ignore_changes = [tags] }
 }
 
 resource "azurerm_storage_account_network_rules" "staticweb" {

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -22,16 +22,10 @@ resource "azurerm_storage_account" "staticweb" {
   lifecycle { ignore_changes = [tags] }
 }
 
-resource "azurerm_storage_container" "staticweb" {
-  name                  = "$web"
-  storage_account_name  = azurerm_storage_account.staticweb.name
-  container_access_type = "private"
-}
-
 resource "azurerm_storage_blob" "staticweb" {
   name                   = "index.html"
   storage_account_name   = azurerm_storage_account.staticweb.name
-  storage_container_name = azurerm_storage_container.staticweb.name
+  storage_container_name = "$web"
   type                   = "Block"
   content_type           = "text/html"
   source_content         = local.staticweb_index_file_content

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -43,8 +43,7 @@ resource "azurerm_storage_account_network_rules" "staticweb" {
   ]
 }
 
-# Assign the role to allow uploading content to the storage account
-# Needed for uploading certificates
+# Assign the "Storage Blob Data Contributor" role needed for uploading certificates to the storage account
 resource "azurerm_role_assignment" "stgwriter" {
   scope                = azurerm_storage_account.staticweb.id
   role_definition_name = "Storage Blob Data Contributor"

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -41,8 +41,8 @@ resource "azurerm_storage_account_network_rules" "staticweb" {
   resource_group_name  = var.resource_group_name
   storage_account_name = azurerm_storage_account.staticweb.name
 
-  default_action = "Deny"
   bypass         = ["AzureServices"]
+  default_action = "Deny"
 
   depends_on = [
     azurerm_storage_blob.staticweb

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -30,6 +30,10 @@ resource "azurerm_storage_blob" "staticweb" {
   content_type           = "text/html"
   source_content         = local.staticweb_index_file_content
 
+  tags = {
+    tre_id = var.tre_id
+  }
+
   lifecycle { ignore_changes = [tags] }
 }
 

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -38,6 +38,7 @@ resource "azurerm_storage_blob" "staticweb" {
   storage_account_name   = azurerm_storage_account.staticweb.name
   storage_container_name = azurerm_storage_container.staticweb.name
   type                   = "Block"
+  content_type           = "text/html"
   source_content         = local.staticweb_index_file_content
 }
 

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -30,6 +30,9 @@ resource "azurerm_storage_blob" "staticweb" {
   content_type           = "text/html"
   source_content         = local.staticweb_index_file_content
 
+  # The storage will not be accessible to Terraform once the network security rules have been applied
+  # Luckily, this blob will not have to be changed once it is created
+  # Hence, we tell Terraform not to touch it again:
   lifecycle {
     ignore_changes = all
   }
@@ -42,6 +45,8 @@ resource "azurerm_storage_account_network_rules" "staticweb" {
   bypass         = ["AzureServices"]
   default_action = "Deny"
 
+  # The rule has to be applied AFTER the blob has been created
+  # Otherwise Terraform will fail as it cannot access the storage account
   depends_on = [
     azurerm_storage_blob.staticweb
   ]

--- a/templates/core/terraform/appgateway/staticweb.tf
+++ b/templates/core/terraform/appgateway/staticweb.tf
@@ -20,36 +20,11 @@ resource "azurerm_storage_account" "staticweb" {
   }
 
   lifecycle { ignore_changes = [tags] }
-}
 
-resource "azurerm_storage_blob" "staticweb" {
-  name                   = "index.html"
-  storage_account_name   = azurerm_storage_account.staticweb.name
-  storage_container_name = "$web"
-  type                   = "Block"
-  content_type           = "text/html"
-  source_content         = local.staticweb_index_file_content
-
-  # The storage will not be accessible to Terraform once the network security rules have been applied
-  # Luckily, this blob will not have to be changed once it is created
-  # Hence, we tell Terraform not to touch it again:
-  lifecycle {
-    ignore_changes = all
+  network_rules {
+    bypass         = ["AzureServices"]
+    default_action = "Deny"
   }
-}
-
-resource "azurerm_storage_account_network_rules" "staticweb" {
-  resource_group_name  = var.resource_group_name
-  storage_account_name = azurerm_storage_account.staticweb.name
-
-  bypass         = ["AzureServices"]
-  default_action = "Deny"
-
-  # The rule has to be applied AFTER the blob has been created
-  # Otherwise Terraform will fail as it cannot access the storage account
-  depends_on = [
-    azurerm_storage_blob.staticweb
-  ]
 }
 
 # Assign the "Storage Blob Data Contributor" role needed for uploading certificates to the storage account

--- a/templates/core/terraform/scripts/letsencrypt.sh
+++ b/templates/core/terraform/scripts/letsencrypt.sh
@@ -10,6 +10,8 @@ fi
 
 IPADDR=$(curl ipecho.net/plain; echo)
 
+# The storage account is protected by network rules
+# The rules need to be temporarily lifted so that the certificate can be uploaded
 echo "Creating network rule on storage account ${STORAGE_ACCOUNT} for $IPADDR"
 az storage account network-rule add --account-name "${STORAGE_ACCOUNT}" --ip-address $IPADDR
 echo "Waiting for network rule to take effect"

--- a/templates/core/terraform/scripts/letsencrypt.sh
+++ b/templates/core/terraform/scripts/letsencrypt.sh
@@ -3,70 +3,6 @@ set -e
 
 script_dir=$(realpath $(dirname "${BASH_SOURCE[0]}"))
 
-if [[ -z ${STORAGE_ACCOUNT} ]]; then
-  echo "STORAGE_ACCOUNT not set"
-  exit 1
-fi
-
-IPADDR=$(curl ipecho.net/plain; echo)
-
-echo "Creating network rule on storage account ${STORAGE_ACCOUNT} for $IPADDR"
-az storage account network-rule add --account-name "${STORAGE_ACCOUNT}" --ip-address $IPADDR
-echo "Waiting for network rule to take effect"
-sleep 30s
-echo "Created network rule on storage account"
-
-staticenabled=$(az storage blob service-properties show -o tsv \
-    --account-name "${STORAGE_ACCOUNT}" \
-    --auth-mode login \
-    --query 'staticWebsite.enabled')
-
-if [[ ${staticenabled} = 'false' ]]; then
-    # Enable static-website hosting
-    echo "Enabling static Web hosting on storage account"
-    az storage blob service-properties update \
-        --account-name "${STORAGE_ACCOUNT}" \
-        --auth-mode login \
-        --static-website \
-        --index-document 'index.html'
-else
-    echo "Static Web hosting already enabled on storage account"
-fi
-
-# Create the default index.html page
-cat << EOF > index.html
-<!DOCTYPE html>
-<html lang="en" xmlns="http://www.w3.org/1999/xhtml"><head><meta charset="utf-8"/><title></title></head><body></body></html>
-EOF
-
-echo "Checking for index.html file in storage account"
-
-isindex=$(az storage blob list -o json \
-    --account-name "${STORAGE_ACCOUNT}" \
-    --auth-mode login \
-    --container-name '$web' \
-    --query "[?name=='index.html'].name" \
-    | jq 'length')
-
-if [[ ${isindex} -lt 1 ]]; then
-    echo "Uploading index.html file"
-    # Upload the default index.html page
-    az storage blob upload \
-        --account-name "${STORAGE_ACCOUNT}" \
-        --auth-mode login \
-        --container-name '$web' \
-        --file index.html \
-        --name index.html \
-        --no-progress \
-        --only-show-errors
-
-    # Wait a bit for the App gateway health probe to notice.
-    echo "Waiting 30s for health probe"
-    sleep 30s
-else
-    echo "index.html already present"
-fi
-
 ledir=$(pwd)/letsencrypt
 
 mkdir -p "${ledir}/logs"
@@ -117,7 +53,3 @@ else
         --cert-file "${CERT_DIR}/aci.pfx" \
         --cert-password "${CERT_PASSWORD}"
 fi
-
-echo "Removing network rule on storage account"
-az storage account network-rule remove --account-name ${STORAGE_ACCOUNT} --ip-address ${IPADDR}
-echo "Removed network rule on storage account"

--- a/templates/core/terraform/scripts/letsencrypt.sh
+++ b/templates/core/terraform/scripts/letsencrypt.sh
@@ -3,6 +3,19 @@ set -e
 
 script_dir=$(realpath $(dirname "${BASH_SOURCE[0]}"))
 
+if [[ -z ${STORAGE_ACCOUNT} ]]; then
+  echo "STORAGE_ACCOUNT not set"
+  exit 1
+fi
+
+IPADDR=$(curl ipecho.net/plain; echo)
+
+echo "Creating network rule on storage account ${STORAGE_ACCOUNT} for $IPADDR"
+az storage account network-rule add --account-name "${STORAGE_ACCOUNT}" --ip-address $IPADDR
+echo "Waiting for network rule to take effect"
+sleep 30s
+echo "Created network rule on storage account"
+
 ledir=$(pwd)/letsencrypt
 
 mkdir -p "${ledir}/logs"
@@ -44,7 +57,6 @@ if [[ -n ${KEYVAULT} ]]; then
         --gateway-name "${APPLICATION_GATEWAY}" \
         --name 'cert-primary' \
         --key-vault-secret-id "${sid}"
-
 else
     az network application-gateway ssl-cert update \
         --resource-group "${RESOURCE_GROUP}" \
@@ -53,3 +65,7 @@ else
         --cert-file "${CERT_DIR}/aci.pfx" \
         --cert-password "${CERT_PASSWORD}"
 fi
+
+echo "Removing network rule on storage account"
+az storage account network-rule remove --account-name ${STORAGE_ACCOUNT} --ip-address ${IPADDR}
+echo "Removed network rule on storage account"

--- a/templates/core/terraform/scripts/letsencrypt.sh
+++ b/templates/core/terraform/scripts/letsencrypt.sh
@@ -11,12 +11,46 @@ fi
 IPADDR=$(curl ipecho.net/plain; echo)
 
 # The storage account is protected by network rules
-# The rules need to be temporarily lifted so that the certificate can be uploaded
+# The rules need to be temporarily lifted so that the index.html file, if required, and certificate can be uploaded
 echo "Creating network rule on storage account ${STORAGE_ACCOUNT} for $IPADDR"
 az storage account network-rule add --account-name "${STORAGE_ACCOUNT}" --ip-address $IPADDR
 echo "Waiting for network rule to take effect"
 sleep 30s
 echo "Created network rule on storage account"
+
+echo "Checking for index.html file in storage account"
+
+# Create the default index.html page
+cat << EOF > index.html
+<!DOCTYPE html>
+<html lang="en" xmlns="http://www.w3.org/1999/xhtml"><head><meta charset="utf-8"/><title></title></head><body></body></html>
+EOF
+
+indexExists=$(az storage blob list -o json \
+    --account-name "${STORAGE_ACCOUNT}" \
+    --auth-mode login \
+    --container-name '$web' \
+    --query "[?name=='index.html'].name" \
+    | jq 'length')
+
+if [[ ${indexExists} -lt 1 ]]; then
+    echo "Uploading index.html file"
+
+    az storage blob upload \
+        --account-name "${STORAGE_ACCOUNT}" \
+        --auth-mode login \
+        --container-name '$web' \
+        --file index.html \
+        --name index.html \
+        --no-progress \
+        --only-show-errors
+
+    # Wait a bit for the App Gateway health probe to notice
+    echo "Waiting 30s for health probe"
+    sleep 30s
+else
+    echo "index.html already present"
+fi
 
 ledir=$(pwd)/letsencrypt
 

